src/ast/nodes/shared/Node.ts

```diff
@@ -86,44 +81,65 @@ export interface ExpressionNode extends ExpressionEntity, Node {}
 export class NodeBase implements ExpressionNode {
 	type: string;
 	keys: string[];
-	included: boolean;
 	scope: Scope;
 	start: number;
 	end: number;
 	module: Module;
-	parent: Node | { type?: string };
-	__enhanced: boolean;
+	parent: Node | { type: string; module: Module };
+	included: boolean;
 
-	constructor() {
-		this.keys = [];
+	constructor(
+		esTreeNode: GenericEsTreeNode,
+		// we need to pass down the node constructors to avoid a circular dependency
+		nodeConstructors: { [p: string]: typeof NodeBase },
+		parent: Node | { type: string; module: Module },
+		parentScope: Scope,
+		preventNewScope: boolean
+	) {
+		this.keys = keys[esTreeNode.type] || getAndCreateKeys(esTreeNode);
+		this.parent = parent;
+		this.module = parent.module;
+		this.createScope(parentScope, preventNewScope);
+		this.parseNode(esTreeNode, nodeConstructors);
+		this.initialise();
+		this.module.magicString.addSourcemapLocation(this.start);
+		this.module.magicString.addSourcemapLocation(this.end);
```

Comment:
At the moment, this would not make a difference. However I actively decided to reset `included` on each node individually so that if we want to reuse an existing AST, we just need to call `initialise` on each node to get it back to its state before tree-shaking.

If you feel this is a premature change as this logic is currently not used, I can of course move this here.